cvefeed/diff.go

```diff
@@ -0,0 +1,301 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cvefeed
+
+import (
+	"encoding/json"
+	"math/bits"
+
+	"github.com/facebookincubator/nvdtools/cvefeed/nvd"
+	"github.com/facebookincubator/nvdtools/cvefeed/nvd/schema"
+)
+
+type bag map[string]interface{}
+
+// ChunkKind is the type of chunks produced by a diff.
+type ChunkKind string
+
+const (
+	// ChunkDescription indicates a difference in the description of a
+	// vulnerability.
+	ChunkDescription ChunkKind = "description"
+	// ChunkScore indicates a difference in the score of a vulnerability.
+	ChunkScore = "score"
+)
+
+type chunk uint32
+
+const (
+	chunkDescriptionShift = iota
+	chunkScoreShift
+	chunkMaxShift
+)
+
+const (
+	chunkDescription chunk = 1 << iota
+	chunkScore
+)
+
+var chunkKind = [chunkMaxShift]ChunkKind{
+	ChunkDescription,
+	ChunkScore,
+}
+
+func chunkShiftByName(kind ChunkKind) int {
+	for i, v := range chunkKind {
+		if v == kind {
+			return i
+		}
+	}
+	return chunkMaxShift
+}
+
+type diffEntry struct {
+	id   string
+	bits chunk
+}
+
+type diffFeed struct {
+	name string
+	dict Dictionary
+}
+
+type diffInput struct {
+	a, b diffFeed
+}
+
+type diff struct {
+	diffInput
+}
+
+func newDiff(i diffInput) *diff {
+	return &diff{
+		diffInput: i,
+	}
+}
+
+// DiffStats is the result of a diff.
+type DiffStats struct {
+	diff *diff // back pointer to the diff these stats are for
+
+	numVulnsA, numVulnsB int
+
+	aNotb     []string // ids of vulns that are in a but not in b
+	bNota     []string // ids of vulns that are in b but not in a
+	entries   []diffEntry
+	bitCounts [chunkMaxShift]int
+}
+
+// NumVulnsA returns the vulnerability in A (the first input to Diff).
+func (s *DiffStats) NumVulnsA() int {
+	return s.numVulnsA
+}
+
+// NumVulnsB returns the vulnerability in A (the first input to Diff).
+func (s *DiffStats) NumVulnsB() int {
+	return s.numVulnsB
+}
+
+// VulnsANotB returns the vulnerabilities that are A (the first input to Diff) but
+// are not in B (the second input to Diff).
+func (s *DiffStats) VulnsANotB() []string {
+	return s.aNotb
+}
+
+// NumVulnsANotB returns the numbers of vulnerabilities that are A (the first input
+// to Diff) but are not in B (the second input to Diff).
+func (s *DiffStats) NumVulnsANotB() int {
+	return len(s.aNotb)
+}
+
+// VulnsBNotA returns the vulnerabilities that are A (the first input to Diff) but
+// are not in B (the second input to Diff).
+func (s *DiffStats) VulnsBNotA() []string {
+	return s.bNota
+}
+
+// NumVulnsBNotA returns the numbers of vulnerabilities that are B (the second input
+// to Diff) but are not in A (the first input to Diff).
+func (s *DiffStats) NumVulnsBNotA() int {
+	return len(s.bNota)
+}
+
+// NumDiffVulns returns the number of vulnerability that are in both A and B but
+// are different (eg. different description, score, ...).
+func (s *DiffStats) NumDiffVulns() int {
+	return len(s.entries)
+}
```

Comment:
Personal opinion, feel free to ignore: I'm not a big fan of gettters/setters in Go. I'd use the fields directly and only introduce these methods once some polymorphism is really needed (preferrably  never).

Reply:
Indeed I wondered about that as well but because I started that way, I continued. I think some of them may create enough "value" to the API user that then doesn't have to understand the internals of DiffStat. Meh, a bit too much work to change now but happy if one of us want to change it later.